### PR TITLE
plugins: add plugin environment validation (CRAFT-611)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Plugins may require certain environment elements to be present in
order to build a part, regardless of how these elements were installed
on the system. For example, a compiler may have been installed from a
deb package, a snap, built from sources or even built by a different
part. This commit adds plugin environment validators, allowing plugins
to verify if the system environment is correct before building a part.

**Note to reviewers**: see the integration test for use cases.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Comment by cmatsuoka** [comment]
Environment variable handling needs to be reworked, the different parts can't be handled in a single dictionary (due to the `PATH=$PATH:...` scenario).

File: craft_parts/executor/environment.py [hunk 1]
```diff
@@ -80,6 +80,33 @@ def generate_part_environment(
...
+    :return: The environment to use when executing the step.
+    """
+    # Craft parts' say.
+    build_environment = _basic_environment_for_part(part=part, step_info=step_info)
+
+    step = step_info.step
+
+    # Plugin's say.
+    if step == Step.BUILD:
+        build_environment.update(plugin.get_build_environment())
```
Comment (cmatsuoka) [inline craft_parts/executor/environment.py:101]: We can't handle env vars using a single dictionary.

File: craft_parts/errors.py [hunk 2]
```diff
@@ -395,6 +395,20 @@ def __init__(self, *, part_name: str, conflicting_files: List[str]):
         super().__init__(brief=brief, details=details)
 
 
+class PluginEnvironmentValidationError(PartsError):
+    """Plugin build script failed at runtime.
```
Comment (sergiusens) [inline craft_parts/errors.py:399]:
```suggestion
    """Plugin validation script failed at runtime.
```

File: tests/integration/plugins/test_validate_environment.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        """
+        try:
+            output = self._execute("mytool")
+            if output.strip() != "ok":
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason="mytool is expected to print ok",
+                )
+        except subprocess.CalledProcessError as err:
+            if err.returncode != 127:
```
Comment (sergiusens) [inline tests/integration/plugins/test_validate_environment.py:80]: code comment for 127 please

File: tests/integration/plugins/test_validate_environment.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    )
+
+    # plugins act on the build step
+    actions = lf.plan(Step.BUILD)
+
+    with lf.action_executor() as exe:
+        exe.execute(actions)
+
+
+def test_validate_plugin_created_in_part(new_dir):
```
Comment (sergiusens) [inline tests/integration/plugins/test_validate_environment.py:165]:
```suggestion
def test_validate_plugin_satisfied_with_part(new_dir):
```

File: craft_parts/plugins/validator.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            supplied by another part when the method is called from the execution
+            prologue. If the validation fails and list is empty, the error is
+            final (the missing elements can't be supplied by a different part).
+            The plugin may require a specific part name as a hint that the
+            part will attempt to supply missing environment elements.
+
+        :raises PluginEnvironmentValidationError: If the environment is invalid.
+        """
+
+    def _execute(self, cmd: str) -> str:
```
Comment (sergiusens) [inline craft_parts/plugins/validator.py:65]: if pylint were enabled, the recommendation would be to make this public :-)

**Comment by sergiusens** [comment]
Looks pretty good, I will do a local pass tomorrow

File: craft_parts/errors.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/errors.py:399]: Copy-and-paste strikes again.

File: tests/integration/plugins/test_validate_environment.py (hunk 3 above)
Comment (cmatsuoka) [reply tests/integration/plugins/test_validate_environment.py:80]: I'll use a constant to replace the magic number.

File: tests/integration/plugins/test_validate_environment.py (hunk 4 above)
Comment (cmatsuoka) [reply tests/integration/plugins/test_validate_environment.py:165]: Renaming as suggested.

File: craft_parts/plugins/validator.py (hunk 5 above)
Comment (cmatsuoka) [reply craft_parts/plugins/validator.py:65]: I run pylint locally, it actually doesn't care about this name in particular.

File: craft_parts/executor/part_handler.py [hunk 6]
```diff
@@ -340,12 +341,25 @@ def _run_step(
         :return: If step is Stage or Prime, return a tuple of sets containing
             the step's file and directory artifacts.
         """
+        step_env = generate_step_environment(
+            part=self._part, plugin=self._plugin, step_info=step_info
+        )
+
+        # Validate build environment
```
Comment (sergiusens) [inline craft_parts/executor/part_handler.py:348]: add comment about differences with the one at the executor level

**Review by sergiusens — APPROVED** [review]

File: craft_parts/executor/part_handler.py (hunk 6 above)
Comment (cmatsuoka) [reply craft_parts/executor/part_handler.py]: Added clarification.

File: craft_parts/plugins/validator.py (hunk 5 above)
Comment (cmatsuoka) [reply craft_parts/plugins/validator.py:65]: To clarify: this is a "protected" method implementing an execution helper intended to be used only by plugin validators. It executes a command in the same environment used to run the build step.

**Pull request merged**